communication/src/allocator/zero_copy/tcp.rs

```diff
@@ -35,12 +33,14 @@ pub fn recv_loop<S>(
     worker_offset: usize,
     process: usize,
     remote: usize,
-    mut logger: Option<Logger<CommunicationEvent>>)
+    logger: Option<Logger<CommunicationEventBuilder>>)
 where
     S: Stream,
 {
     // Log the receive thread's start.
-    logger.as_mut().map(|l| l.log(StateEvent { send: false, process, remote, start: true }));
+    if let Some(logger) = logger.as_ref() {
+        logger.log(CommunicationEvent::from(StateEvent { send: false, process, remote, start: true }));
```

Comment:
I've seen a bunch of these, and don't know yet but wonder if `log` should take a `T: From<Foo>` where `Foo` is the type it otherwise expects as its argument.

Reply:
That's hard because we need a T that can be copied into the container (builder), and that supports From for some argument type. I don't think rust could figure out the in-between type because there might be more than one option. This means we have to specify it somewhere manually....

Reply:
Gotcha. One option would be to have a `Logger<T, CB: PushInto<T>>`, which carries the opinions about what you should be pushing in to it. I can't really tell whether it is worse giving up the generality of `PushInto`, as that would do, or giving up the `Into` constraint and asking folks to build the logged types more explicitly.